keanu-project/src/main/java/io/improbable/keanu/algorithms/sampling/ForwardSampler.java

```diff
@@ -0,0 +1,114 @@
+                                        KeanuRandom random) {
+
+        Preconditions.checkArgument(network.getObservedVertices().size() == 0, "Cannot forward sample from a network with observations");
+
+        List<Vertex> fromVertices = fromVariables.stream().map(v -> (Vertex) v).collect(Collectors.toList());
+
+        Set<Vertex> probabilisticSubset = fromVertices.stream().filter(v -> v instanceof Probabilistic).collect(Collectors.toSet());
+        Set<Vertex> nonProbabilisticSubset = fromVertices.stream().filter( v -> !(v instanceof Probabilistic)).collect(Collectors.toSet());
+
+        upstreamOfProbabilisticDoesNotContainProbabilistic(probabilisticSubset);
+
+        ComputableModel graph = new KeanuComputableModel(new HashSet<>(network.getAllVertices()));
+        KeanuProposalApplicationStrategy proposalApplicationStrategy = new KeanuProposalApplicationStrategy(probabilisticSubset);
+
+        return sample(graph, probabilisticSubset, nonProbabilisticSubset, proposalApplicationStrategy, sampleCount, random);
+    }
+
+    private NetworkSamples sample(ComputableModel graph,
+                                        Set<? extends Variable> probabilisticFromVariables,
+                                        Set<? extends Variable> nonProbabilisticFromVariables,
+                                        ProposalApplicationStrategy proposalApplicationStrategy,
+                                        int sampleCount,
+                                        KeanuRandom random) {
+
+        Set<? extends Variable> allSampleFromVariables = Sets.union(probabilisticFromVariables, nonProbabilisticFromVariables);
+        Map<VariableReference, List> samplesByVariable = new HashMap<>();
+
+        for (int sampleNum = 0; sampleNum < sampleCount; sampleNum++) {
+            Proposal probabilisticVariableProposal = PRIOR_PROPOSAL.getProposal(probabilisticFromVariables, random);
+            proposalApplicationStrategy.apply(probabilisticVariableProposal);
+            graph.compute(Collections.EMPTY_MAP, nonProbabilisticFromVariables.stream().map(Variable::getReference).collect(Collectors.toList()));
+            takeSamples(samplesByVariable, allSampleFromVariables);
+        }
+
+        ArrayList<Double> logProb = new ArrayList<>(Collections.nCopies(sampleCount, LOG_PROB_OF_PRIOR));
+
+        return new NetworkSamples(samplesByVariable, logProb, sampleCount);
+    }
+
+    private void takeSamples(Map<VariableReference, List> samples, Set<? extends Variable> fromVariables) {
+        fromVariables.forEach(variable -> addVariableValue(variable, samples));
+    }
+
+    private void addVariableValue(Variable variable, Map<VariableReference, List> samples) {
+        List samplesForVariable = samples.computeIfAbsent(variable.getReference(), v -> new ArrayList<>());
+        samplesForVariable.add(variable.getValue());
+    }
+
+    private void upstreamOfProbabilisticDoesNotContainProbabilistic(Set<Vertex> vertices) {
```

Comment:
This seems overly strict? (and certainly wouldn't work for the model that initially prompted ChrisM's implementation).

Reply:
Relaxed the constraint now that we're happy with conditionals